snapshot/cmd/snapshot-pv-provisioner/snapshot-pv-provisioner.go

```diff
@@ -0,0 +1,302 @@
+		Namespace(v1.NamespaceDefault).
+		Name(snapshot.Spec.SnapshotDataName).
+		Do().Into(&snapshotData)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve VolumeSnapshotData %s: %v", snapshot.Spec.SnapshotDataName, err)
+	}
+	glog.V(3).Infof("restore from VolumeSnapshotData %s", snapshot.Spec.SnapshotDataName)
+
+	pvSrc, labels, err := p.snapshotRestore(snapshot.Spec.SnapshotDataName, snapshotData, options)
+	if err != nil || pvSrc == nil {
+		return nil, fmt.Errorf("failed to create a PV from snapshot %s: %v", snapshotName, err)
+	}
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: options.PVName,
+			Annotations: map[string]string{
+				provisionerIDAnn: p.identity,
+			},
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeReclaimPolicy: options.PersistentVolumeReclaimPolicy,
+			AccessModes:                   options.PVC.Spec.AccessModes,
+			Capacity: v1.ResourceList{
+				v1.ResourceName(v1.ResourceStorage): options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)],
+			},
+			PersistentVolumeSource: *pvSrc,
+		},
+	}
+
+	if len(labels) != 0 {
+		if pv.Labels == nil {
+			pv.Labels = make(map[string]string)
+		}
+		for k, v := range labels {
+			pv.Labels[k] = v
+		}
+	}
+
+	glog.Infof("successfully created Snapshot share %#v", pv)
+
+	return pv, nil
+}
+
+// Delete removes the storage asset that was created by Provision represented
+// by the given PV.
+func (p *snapshotProvisioner) Delete(volume *v1.PersistentVolume) error {
+	ann, ok := volume.Annotations[provisionerIDAnn]
+	if !ok {
+		return errors.New("identity annotation not found on PV")
+	}
+	if ann != p.identity {
+		return &controller.IgnoredError{Reason: "identity annotation on PV does not match ours"}
+	}
+
+	volumeType := crdv1.GetSupportedVolumeFromPVSpec(&volume.Spec)
+	if len(volumeType) == 0 {
+		return fmt.Errorf("unsupported volume type found in PV %#v", *volume)
+	}
+	plugin, ok := volumePlugins[volumeType]
+	if !ok {
+		return fmt.Errorf("%s is not supported volume for %#v", volumeType, *volume)
+	}
+
+	// delete PV
+	return plugin.VolumeDelete(volume)
+}
+
+var (
+	master          = flag.String("master", "", "Master URL")
+	kubeconfig      = flag.String("kubeconfig", "", "Absolute path to the kubeconfig")
+	id              = flag.String("id", "", "Unique provisioner identity")
+	cloudProvider   = flag.String("cloudprovider", "", "aws|gce|openstack")
+	cloudConfigFile = flag.String("cloudconfig", "", "Path to a Cloud config. Only required if cloudprovider is set.")
+	volumePlugins   = make(map[string]volume.VolumePlugin)
+)
+
+func main() {
+	flag.Parse()
+	flag.Set("logtostderr", "true")
+
+	var config *rest.Config
+	var err error
+	if *master != "" || *kubeconfig != "" {
+		config, err = clientcmd.BuildConfigFromFlags(*master, *kubeconfig)
+	} else {
+		config, err = rest.InClusterConfig()
+	}
+	prId := string(uuid.NewUUID())
+	if *id != "" {
+		prId = *id
```

Comment:
We should default to provisionerName instead of generating every start up like https://github.com/kubernetes-incubator/external-storage/blob/master/ceph/rbd/cmd/rbd-provisioner/main.go#L67

this means Delete will succeed from any instance of the provisioner with same provisionername, I assume that is safe.